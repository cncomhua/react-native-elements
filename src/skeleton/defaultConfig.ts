import { View, ViewStyle } from 'react-native';

type defaultLinearGradientPropsReturnParams = {
  colors: string[];
  start: { x: number; y: number };
  end: { x: number; y: number };
};

const defaultProps = {
  ViewComponent: View,
  duration: 1000,
  fluid: false,
  rounded: false,
  height: 20,
  width: '100%',
  containerStyle: {},
  easingType: 'ease',
  linearGradientProps: {},
};

const defaultContainerStyle: ViewStyle = {
  overflow: 'hidden',
  borderRadius: 10,
};

// workaround to pass linear gradient props to View
const createDefaultLinearGradientProps = (
  backgroundColor: string,
  skeletonColor: string
): defaultLinearGradientPropsReturnParams => ({
  colors: [backgroundColor, skeletonColor, backgroundColor],
  start: { x: 0, y: 0 },
  end: { x: 1, y: 0 },
});

export {
  defaultContainerStyle,
  createDefaultLinearGradientProps,
  defaultProps,
};
