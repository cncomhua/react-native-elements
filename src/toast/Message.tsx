import React, {
  useEffect,
  useRef,
  useContext,
  useMemo,
  useCallback,
} from 'react';
import { Animated, Text, StyleSheet } from 'react-native';

import {
  StylePropsWithMessageType,
  ToastContext,
  ToastPosition,
  ToastTypes,
} from './ToastProvider';

import type { VFC } from 'react';
import type { MessageState } from './ToastProvider';
import type { StyleProp, ViewStyle, TextStyle } from 'react-native';

type MessageProps = {
  message: MessageState;
  onHide: () => void;
};

const Message: VFC<MessageProps> = ({ message, onHide }) => {
  const {
    duration,
    position,
    containerMessageStyle,
    textMessageStyle,
  } = useContext(ToastContext);

  const opacity = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    Animated.sequence([
      Animated.timing(opacity, {
        toValue: 1,
        duration: 300,
        useNativeDriver: true,
      }),
      Animated.delay(duration),
      Animated.timing(opacity, {
        toValue: 0,
        duration: 300,
        useNativeDriver: true,
      }),
    ]).start(() => onHide());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const animationContainerStyles = useMemo(
    () => ({
      opacity,
      transform: [
        {
          translateY: opacity.interpolate({
            inputRange: [0, 1],
            outputRange: [position === ToastPosition.top ? -20 : 20, 0],
          }),
        },
      ],
    }),
    [opacity, position]
  );

  const createTypedStyles = useCallback(
    function <T>(styles: StylePropsWithMessageType<T>): StyleProp<T> {
      const localStyles = {
        ...(styles as object),
      } as StylePropsWithMessageType<T>;
      const typedStyles = { ...((localStyles[message.type] ?? {}) as object) };

      for (let key in ToastTypes) {
        delete localStyles[key as keyof StylePropsWithMessageType<T>];
      }

      return Object.keys(typedStyles as object).length > 0
        ? ({
            ...(localStyles as object),
            ...(typedStyles as object),
          } as StyleProp<T>)
        : localStyles;
    },
    [message.type]
  );

  const typedContainerStyles = createTypedStyles<ViewStyle>(
    containerMessageStyle
  );
  const typedTextStyles = createTypedStyles<TextStyle>(textMessageStyle);

  return (
    <Animated.View
      style={StyleSheet.flatten([
        animationContainerStyles,
        styles.messageContainer,
        typedContainerStyles,
      ])}
    >
      <Text style={StyleSheet.flatten([styles.message, typedTextStyles])}>
        {message.text}
      </Text>
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  messageContainer: {
    margin: 10,
    marginBottom: 5,
    padding: 10,
    borderRadius: 4,
    shadowColor: 'black',
    shadowOffset: {
      width: 0,
      height: 3,
    },
    shadowOpacity: 0.15,
    shadowRadius: 5,
    elevation: 6,
  },
  message: {},
});

export default Message;
