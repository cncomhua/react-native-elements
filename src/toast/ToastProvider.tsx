import React, { createContext, useState } from 'react';
import { StyleSheet } from 'react-native';

import type { Dispatch, SetStateAction } from 'react';
import type { StyleProp, TextStyle, ViewStyle, TextProps } from 'react-native';
import type { DefaultConfigType } from './config';
import type { RneFunctionComponent } from '../helpers';

import defaultConfig from './config';
import Toast from './Toast';
import { withTheme } from '../config';

export enum ToastTypes {
  info = 'info',
  success = 'success',
  warning = 'warning',
  error = 'error',
}

export enum ToastPosition {
  bottom = 'bottom',
  top = 'top',
}

export type StylePropsWithMessageType<T> = StyleProp<T> &
  {
    [key in ToastTypes]?: StyleProp<T>;
  };

export type MessageState = {
  id: string;
  text: string;
  type: ToastTypes;
};

export type ToastProviderProps = {
  duration?: number;
  maxMessages?: number;
  position?: keyof typeof ToastPosition;
  containerToastStyle?: StyleProp<ViewStyle>;
  containerMessageStyle?: StylePropsWithMessageType<ViewStyle>;
  textMessageStyle?: StylePropsWithMessageType<TextStyle>;
  textMessageProps?: TextProps;
};

type ToastContextType = {
  messages: MessageState[];
  setMessage: Dispatch<SetStateAction<MessageState[]>>;
} & DefaultConfigType;

export const ToastContext = createContext<ToastContextType>({
  messages: [],
  setMessage: () => [],
  duration: defaultConfig.duration,
  maxMessages: defaultConfig.maxMessages,
  position: defaultConfig.position,
  containerToastStyle: defaultConfig.containerToastStyle,
  containerMessageStyle: defaultConfig.containerMessageStyle,
  textMessageStyle: defaultConfig.textMessageStyle,
  textMessageProps: defaultConfig.textMessageProps,
});

const ToastProvider: RneFunctionComponent<ToastProviderProps> = ({
  children,
  duration,
  maxMessages,
  position,
  containerToastStyle,
  containerMessageStyle,
  textMessageStyle,
  textMessageProps,
}) => {
  const {
    duration: defaultDuration,
    maxMessages: defaultMaxMessages,
    position: defaultPosition,
    containerToastStyle: defaultContainerToastStyle,
    containerMessageStyle: defaultContainerMessageStyle,
    textMessageStyle: defaultTextMessageStyle,
    textMessageProps: defaultTextMessageProps,
  } = defaultConfig;

  const [messages, setMessage] = useState<MessageState[]>([]);

  return (
    <ToastContext.Provider
      key={'toast-context-test'}
      value={{
        messages,
        setMessage,
        duration: duration ?? defaultDuration,
        maxMessages: maxMessages ?? defaultMaxMessages,
        position: ToastPosition[position ?? defaultPosition] ?? defaultPosition,
        containerToastStyle: StyleSheet.flatten([
          defaultContainerToastStyle,
          containerToastStyle,
        ]),
        containerMessageStyle: StyleSheet.flatten([
          defaultContainerMessageStyle,
          containerMessageStyle as object,
        ]),
        textMessageStyle: StyleSheet.flatten([
          defaultTextMessageStyle,
          textMessageStyle as object,
        ]),
        textMessageProps: {
          ...defaultTextMessageProps,
          ...textMessageProps,
        },
      }}
    >
      <Toast messages={messages} setMessage={setMessage} />
      {children}
    </ToastContext.Provider>
  );
};

export { ToastProvider };

export default withTheme<ToastProviderProps>(ToastProvider, 'ToastProvider');
