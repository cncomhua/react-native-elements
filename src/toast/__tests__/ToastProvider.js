import React from 'react';
import { create, act } from 'react-test-renderer';
import { ToastProvider, useToast } from '../../';
import defaultConfig from '../config';
import { Button, StyleSheet } from 'react-native';
import { ToastPosition } from '../ToastProvider';
import Message from '../Message';
import { ToastTypes } from '../ToastProvider';

describe('Toast provider component', () => {
  jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');

  let component;

  beforeEach(() => {
    component = null;
  });

  const deleteTypedFields = (obj) => {
    for (let key in ToastTypes) {
      delete obj[key];
    }
    return obj;
  };

  it('should render toast provider properly and match snapshot', () => {
    component = create(<ToastProvider />);
    expect(component.root).toBeTruthy();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should handle messages via toast provider', () => {
    const ButtonComponent = () => {
      const { showMessage } = useToast();

      return (
        <Button
          title={'Press me to show message'}
          onPress={() => showMessage("I'm random message")}
        />
      );
    };

    act(() => {
      component = create(
        <ToastProvider>
          <ButtonComponent />
        </ToastProvider>
      );
    });

    const button = component.root.findByType(Button);

    expect(
      component.root.findByProps({ testID: 'toast-container-test' }).props
        .children.length
    ).toBe(0);

    act(() => {
      button.props.onPress();
    });

    act(() => {
      expect(
        component.root.findByProps({ testID: 'toast-container-test' }).props
          .children.length
      ).toBe(1);
    });

    act(() => jest.runAllTimers());

    act(() => {
      expect(
        component.root.findByProps({ testID: 'toast-container-test' }).props
          .children.length
      ).toBe(0);
    });

    act(() => {
      button.props.onPress();
      button.props.onPress();
    });

    act(() => {
      expect(
        component.root.findByProps({ testID: 'toast-container-test' }).props
          .children.length
      ).toBe(2);
    });
  });

  it('should handle default props', () => {
    component = create(
      <ToastProvider>
        <Message
          message={{ id: 0, text: 'Default message', type: 'info' }}
          onHide={() => null}
        />
      </ToastProvider>
    );

    const toast = component.root.findByProps({
      testID: 'toast-container-test',
    });

    expect(toast.props.style).toEqual(
      StyleSheet.flatten([
        {
          top: defaultConfig.position === ToastPosition.top ? 0 : undefined,
          bottom:
            defaultConfig.position === ToastPosition.bottom ? 0 : undefined,
        },
        defaultConfig.containerToastStyle,
      ])
    );

    const message = component.root.findByProps({
      testID: 'messages-test',
    });

    expect(message.props.style).toMatchObject(
      deleteTypedFields(defaultConfig.containerMessageStyle)
    );

    const textMessage = component.root.findByProps({
      testID: 'messages-text-test',
    });

    expect(textMessage.props.style).toMatchObject({
      ...deleteTypedFields(defaultConfig.textMessageStyle),
      ...defaultConfig.textMessageProps,
    });
  });

  it('should handle external props', () => {
    const position = 'bottom';
    const containerToastStyle = {
      marginTop: 10,
      backgroundColor: 'black',
    };
    const containerMessageStyle = {
      padding: 10,
      borderRadius: 0,
    };
    const textMessageStyle = {
      color: 'red',
      fontSize: 10,
    };

    const textMessageProps = {
      numberOfLines: 2,
      onPress: () => 'hi buddy :)',
    };

    component = create(
      <ToastProvider
        position={position}
        containerToastStyle={containerToastStyle}
        containerMessageStyle={containerMessageStyle}
        textMessageStyle={textMessageStyle}
        textMessageProps={textMessageProps}
      >
        <Message
          message={{ id: 0, text: 'Default message', type: 'info' }}
          onHide={() => null}
        />
      </ToastProvider>
    );

    const toast = component.root.findByProps({
      testID: 'toast-container-test',
    });

    expect(toast.props.style).toEqual(
      StyleSheet.flatten([
        {
          top: position === ToastPosition.top ? 0 : undefined,
          bottom: position !== ToastPosition.top ? 0 : undefined,
        },
        defaultConfig.containerToastStyle,
        containerToastStyle,
      ])
    );

    const message = component.root.findByProps({
      testID: 'messages-test',
    });

    expect(message.props.style).toMatchObject({
      ...deleteTypedFields(defaultConfig.containerMessageStyle),
      ...deleteTypedFields(containerMessageStyle),
    });

    const textMessage = component.root.findByProps({
      testID: 'messages-text-test',
    });

    expect(textMessage.props).toMatchObject({
      style: {
        ...deleteTypedFields(defaultConfig.textMessageStyle),
        ...deleteTypedFields(textMessageStyle),
      },
      ...defaultConfig.textMessageProps,
      ...textMessageProps,
    });
  });
});
