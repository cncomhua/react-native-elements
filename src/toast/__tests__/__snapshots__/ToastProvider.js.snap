// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast provider component should render toast provider properly and match snapshot 1`] = `
<View
  style={
    Object {
      "bottom": undefined,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 99999,
    }
  }
  testID="toast-container-test"
/>
`;
